ambry-clustermap/src/main/java/com.github.ambry.clustermap/RecoveryTestClusterAgentsFactory.java

```diff
@@ -0,0 +1,131 @@
+
+import com.github.ambry.config.ClusterMapConfig;
+import com.github.ambry.server.AmbryHealthReport;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import org.json.JSONException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A factory class to construct {@link RecoveryTestClusterManager} and a no op {@link ClusterParticipant}.
+ * Only one instance of each type of objects will ever be created by this factory.
+ */
+public class RecoveryTestClusterAgentsFactory implements ClusterAgentsFactory {
+  private static final Logger logger = LoggerFactory.getLogger(RecoveryTestClusterAgentsFactory.class);
+  private final StaticClusterAgentsFactory staticClusterAgentsFactory;
+  private final HelixClusterAgentsFactory helixClusterAgentsFactory;
+  private final AtomicReference<RecoveryTestClusterManager> recoveryTestClusterManagerRef = new AtomicReference<>();
+  private final AtomicReference<ClusterParticipant> clusterParticipantRef = new AtomicReference<>();
+
+  /**
+   * Create an instance of this class.
+   * @param clusterMapConfig the {@link ClusterMapConfig} to use.
+   * @param hardwareLayoutFilePath the path to the hardware layout file.
+   * @param partitionLayoutFilePath the path to the partition layout file.
+   * @throws JSONException if there is an exception parsing the layout files.
+   * @throws IOException if there is an IO error accessing or reading the layout files.
+   */
+  public RecoveryTestClusterAgentsFactory(ClusterMapConfig clusterMapConfig, String hardwareLayoutFilePath,
+      String partitionLayoutFilePath) throws JSONException {
+    PartitionLayout partitionLayout =
+        new PartitionLayout(new HardwareLayout(clusterMapConfig.clustermapRecoveryTestHardwareLayout, clusterMapConfig),
+            clusterMapConfig.clustermapRecoveryTestPartitionLayout, clusterMapConfig);
+    staticClusterAgentsFactory = new StaticClusterAgentsFactory(clusterMapConfig, partitionLayout);
+    helixClusterAgentsFactory =
+        new HelixClusterAgentsFactory(clusterMapConfig, staticClusterAgentsFactory.getMetricRegistry());
+  }
+
+  /**
+   * Create and return a {@link RecoveryTestClusterManager}.
+   * @return the constructed {@link RecoveryTestClusterManager}.
+   * @throws Exception if constructing the underlying {@link StaticClusterManager} or the {@link HelixClusterManager}
+   * throws an Exception.
+   */
+  @Override
+  public RecoveryTestClusterManager getClusterMap() throws IOException {
+    if (recoveryTestClusterManagerRef.get() == null) {
+      StaticClusterManager staticClusterManager = staticClusterAgentsFactory.getClusterMap();
+      HelixClusterManager helixClusterManager = null;
+      try {
+        helixClusterManager = helixClusterAgentsFactory.getClusterMap();
+      } catch (Exception e) {
+        logger.error("Helix cluster manager instantiation failed with exception", e);
```

Comment:
If there is an exception when getting helix cluster manager, should we terminate the startup?

Reply:
done.